perses/app/fah_generator.py

```diff
@@ -0,0 +1,347 @@
+                             'lambda_angles': x,
+                             'lambda_torsions': x}
+
+
+def make_neq_integrator(nsteps_eq=250000, nsteps_neq=250000, neq_splitting='V R H O R V', timestep=4.0 * unit.femtosecond, alchemical_functions = DEFAULT_ALCHEMICAL_FUNCTIONS, **kwargs):
+    """
+    generate an openmmtools.integrators.PeriodicNonequilibriumIntegrator
+
+    arguments
+        nsteps_eq : int, default=1000
+            Number of equilibration steps to dwell within lambda = 0 or 1 when reached
+        nsteps_neq : int, default=1000
+            Number of nonequilibrium switching steps for 0->1 and 1->0 switches
+        neq_splitting : str, default='V R H O R V'
+            Sequence of "R", "V", "O" (and optionally "{", "}", "V0", "V1", ...) substeps to be executed each timestep.
+            "H" increments the global parameter `lambda` by 1/nsteps_neq for each step and accumulates protocol work.
+        timestep : int, default=4.0 * unit.femtosecond
+            integrator timestep
+        **kwargs :
+            miscellaneous arguments for openmmtools.integrators.LangevinIntegrator
+
+    returns
+        integrator : openmmtools.integrators.PeriodicNonequilibriumIntegrator
+
+    """
+    from openmmtools.integrators import PeriodicNonequilibriumIntegrator
+    #from copy import deepcopy
+    #integrator_kwargs = deepcopy(setup_options_dict)
+    integrator = PeriodicNonequilibriumIntegrator(alchemical_functions, nsteps_eq, nsteps_neq, neq_splitting,timestep=timestep)
+    return integrator
+
+def relax_structure(temperature, system, positions, nminimize=100, nequil = 4, n_steps_per_iteration=250):
+    """
+    arguments
+        temperature : simtk.unit.Quantity with units compatible with kelvin
+            temperature of simulation
+        system : openmm.System
+            system object for simulation
+        positions : simtk.unit.Quantity of shape (natoms,3) with units compatible with nanometers
+            Positions of the atoms in the system
+        nminimize : int
+            number of minimization steps
+        nequil : int
+            number of equilibration applications
+        n_steps_per_iteration : int
+            numper of steps per nequil
+
+    return
+        state : openmm.State
+            state of simulation (getEnergy=True, getForces=True, getPositions=True, getVelocities=True, getParameters=True)
+    """
+
+    from perses.dispersed.feptasks import minimize
+    from openmmtools.integrators import LangevinIntegrator
+    import tqdm
+    _logger.info(f'Starting to relax')
+    integrator = LangevinIntegrator(temperature = temperature)
+    context = openmm.Context(system, integrator)
+    context.setPeriodicBoxVectors(*system.getDefaultPeriodicBoxVectors())
+    context.setPositions(positions)
+
+    _logger.info(f'Starting to minimise')
+    # Minimize
+    for iteration in tqdm.tqdm(range(nminimize)):
+        openmm.LocalEnergyMinimizer.minimize(context, 0.0, 1)
```

Comment:
It may be better to just do one minimization that doesn’t set a step limit or force norm threshold. I realize that LBFGS doesn’t do its thing if you don’t give it enough steps---it may not work right if we run one step at a time.

Reply:
I'd change this from 
```python
for iteration in tqdm.tqdm(range(nminimize)):
        openmm.LocalEnergyMinimizer.minimize(context, 0.0, 1)
```
to
```python
openmm.LocalEnergyMinimizer.minimize(context)
```